### PR TITLE
smt2 parser: fix sign extension

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: e31bf9a533d59844b52aa90d74402f5ef6c98b75).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94048272
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/solvers/smt2/smt2_parser.cpp [hunk 1]
```diff
@@ -781,11 +781,14 @@ exprt smt2_parsert::function_application()
           else if(id=="sign_extend")
           {
             auto width=to_unsignedbv_type(op[0].type()).get_width();
-            signedbv_typet signed_type(width+index);
+            signedbv_typet small_signed_type(width);
+            signedbv_typet large_signed_type(width + index);
```
Comment (romainbrenguier) [inline src/solvers/smt2/smt2_parser.cpp]: both const

**Review by romainbrenguier — APPROVED** [review]

File: src/solvers/smt2/smt2_parser.cpp [hunk 2]
```diff
@@ -781,11 +781,14 @@ exprt smt2_parsert::function_application()
...
             unsignedbv_typet unsigned_type(width+index);
 
             return typecast_exprt(
-              typecast_exprt(op[0], signed_type), unsigned_type);
+              typecast_exprt(
+                typecast_exprt(op[0], small_signed_type), large_signed_type),
```
Comment (allredj) [inline src/solvers/smt2/smt2_parser.cpp:792]: Not sure why we apply the two casts here (small and large). Could you maybe add a comment?

**Review by allredj — COMMENTED** [review]
Comment request. Otherwise looks fine.

File: src/solvers/smt2/smt2_parser.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/smt2/smt2_parser.cpp]: done

File: src/solvers/smt2/smt2_parser.cpp (hunk 2 above)
Comment (kroening) [reply src/solvers/smt2/smt2_parser.cpp:792]: done

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a0164a8dc50ff99405ef4fe0509efd135c718b03).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94355292

**Pull request merged**